File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.java
package expo.modules.barcodescanner;

import android.content.Context;

import org.unimodules.core.ModuleRegistry;
import org.unimodules.core.ViewManager;
import org.unimodules.core.interfaces.ExpoProp;

import java.util.List;
import java.util.ArrayList;

import expo.modules.interfaces.barcodescanner.BarCodeScannerSettings;

public class BarCodeScannerViewManager extends ViewManager<BarCodeScannerView> {
  public enum Events {
    EVENT_ON_BAR_CODE_SCANNED("onBarCodeScanned");

    private final String mName;

    Events(final String name) {
      mName = name;
    }

    @Override
    public String toString() {
      return mName;
    }
  }

  private static final String TAG = "ExpoBarCodeScannerView";
  private ModuleRegistry mModuleRegistry;

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mModuleRegistry = moduleRegistry;
  }

  @Override
  public String getName() {
    return TAG;
  }

  @Override
  public BarCodeScannerView createViewInstance(Context context) {
    return new BarCodeScannerView(context, mModuleRegistry);
  }

  @Override
  public ViewManagerType getViewManagerType() {
    return ViewManagerType.GROUP;
  }

  @Override
  public List<String> getExportedEventNames() {
    List<String> eventNames = new ArrayList<>(Events.values().length);
    for(Events event : Events.values()) {
      eventNames.add(event.toString());
    }
    return eventNames;
  }

  @ExpoProp(name = "type")
  public void setType(BarCodeScannerView view, int type) {
    view.setCameraType(type);
  }

  @ExpoProp(name = "barCodeTypes")
  public void setBarCodeTypes(BarCodeScannerView view, final ArrayList<Double> barCodeTypes) {
    if (barCodeTypes == null) {
      return;
    }

    BarCodeScannerSettings settings = new BarCodeScannerSettings() {{
      putTypes(barCodeTypes);
    }};
    view.setBarCodeScannerSettings(settings);
  }
}

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.kt
package expo.modules.barcodescanner

import android.content.Context
import expo.modules.core.ModuleRegistry
import expo.modules.core.ModuleRegistryDelegate
import expo.modules.core.ViewManager
import expo.modules.core.interfaces.ExpoProp
import expo.modules.interfaces.barcodescanner.BarCodeScannerSettings
import java.util.*

class BarCodeScannerViewManager(
  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
) : ViewManager<BarCodeScannerView>() {

  enum class Events(private val mName: String) {
    EVENT_ON_BAR_CODE_SCANNED("onBarCodeScanned");

    override fun toString() = mName
  }

  override fun getName() = TAG

  override fun onCreate(moduleRegistry: ModuleRegistry) {
    moduleRegistryDelegate.onCreate(moduleRegistry)
  }

  override fun createViewInstance(context: Context) =
    BarCodeScannerView(context, moduleRegistryDelegate)

  override fun getViewManagerType() = ViewManagerType.GROUP

  override fun getExportedEventNames() =
    Events.values().map { it.toString() }

  @ExpoProp(name = "type")
  fun setType(view: BarCodeScannerView, type: Int) {
    view.setCameraType(type)
  }

  @ExpoProp(name = "barCodeTypes")
  fun setBarCodeTypes(view: BarCodeScannerView, barCodeTypes: ArrayList<Double?>?) {
    if (barCodeTypes != null) {
      val settings = BarCodeScannerSettings().apply {
        putTypes(barCodeTypes)
      }
      view.setBarCodeScannerSettings(settings)
    }
  }

  companion object {
    private const val TAG = "ExpoBarCodeScannerView"
  }
}
